base/compiler/ssair/passes.jl

```diff
@@ -1,3 +1,416 @@
+            continue
+        end
+        found_def = true
+        break
+    end
+    found_def ? (def, defidx) : nothing
+end
+
+is_tuple_call(ir, def) = isa(def, Expr) && is_known_call(def, tuple, ir, ir.mod)
+
+function process_immutable_preserve(new_preserves, compact, def)
+    for arg in (isexpr(def, :new) ? def.args : def.args[2:end])
+        if !isbits(compact_exprtype(compact, arg))
+            push!(new_preserves, arg)
+        end
+    end
+end
+
+function getfield_elim_pass!(ir::IRCode, domtree)
+    compact = IncrementalCompact(ir)
+    insertions = Vector{Any}()
+    defuses = IdDict{Int, Tuple{IdSet{Int}, SSADefUse}}()
+    for (idx, stmt) in compact
+        isa(stmt, Expr) || continue
+        is_getfield = false
+        is_ccall = false
+        # Step 1: Check whether the statement we're looking at is a getfield/setfield!
+        if is_known_call(stmt, setfield!, ir, ir.mod)
+            is_setfield = true
+        elseif is_known_call(stmt, getfield, ir, ir.mod)
+            is_getfield = true
+        elseif isexpr(stmt, :foreigncall)
+            nccallargs = stmt.args[5]
+            new_preserves = Any[]
+            old_preserves = stmt.args[(6+nccallargs):end]
+            for (pidx, preserved_arg) in enumerate(old_preserves)
+                intermediaries = IdSet()
+                isa(preserved_arg, SSAValue) || continue
+                def = walk_to_def(compact, preserved_arg, intermediaries, false)
+                def !== nothing || continue
+                (def, defidx) = def
+                if is_tuple_call(ir, def)
+                    process_immutable_preserve(new_preserves, compact, def)
+                    old_preserves[pidx] = nothing
+                    continue
+                elseif isexpr(def, :new)
+                    typ = def.typ
+                    if isa(typ, UnionAll)
+                        typ = unwrap_unionall(typ)
+                    end
+                    if !typ.mutable
+                        process_immutable_preserve(new_preserves, compact, def)
+                        old_preserves[pidx] = nothing
+                        continue
+                    end
+                else
+                    continue
+                end
+                mid, defuse = get!(defuses, defidx, (IdSet{Int}(), SSADefUse()))
+                push!(defuse.ccall_preserve_uses, idx)
+                union!(mid, intermediaries)
+                continue
+            end
+            if !isempty(new_preserves)
+                old_preserves = filter(ssa->ssa !== nothing, old_preserves)
+                new_expr = Expr(:foreigncall, stmt.args[1:(6+nccallargs-1)]...,
+                    old_preserves..., new_preserves...)
+                new_expr.typ = stmt.typ
+                compact[idx] = new_expr
+            end
+            continue
+        else
+            continue
+        end
+        ## Normalize the field argument to getfield/setfield
+        field = stmt.args[3]
```

Comment:
should always be asking `expr_type` for the type, not open-coding a random subset of it